src/frontend/src/utils/condition.rs

```diff
@@ -676,6 +754,139 @@ impl fmt::Debug for ConditionDisplay<'_> {
     }
 }
 
+mod mismatch {
+    use std::ops::Bound;
+
+    use risingwave_common::error::{ErrorCode, RwError};
+    use risingwave_common::types::{DataType, ScalarImpl};
+    use risingwave_pb::expr::expr_node::Type;
+
+    use crate::expr::{Expr, ExprImpl};
+
+    enum ShrinkResult {
+        OutUpperBound,
+        OutLowerBound,
+        InRange(ExprImpl),
+    }
+
+    /// return None indicates that the expression is out of range
+    pub fn cast_mismatch_eq(
+        const_expr: ExprImpl,
+        target: DataType,
+    ) -> Result<Option<ExprImpl>, RwError> {
+        match (const_expr.return_type(), &target) {
+            (DataType::Int64, DataType::Int32)
+            | (DataType::Int64, DataType::Int16)
+            | (DataType::Int32, DataType::Int16) => match shrink_int(const_expr, target)? {
+                ShrinkResult::OutUpperBound | ShrinkResult::OutLowerBound => Ok(None),
+                ShrinkResult::InRange(expr) => Ok(Some(expr)),
+            },
+            _ => Err(ErrorCode::BindError(format!(
+                "Cannot cast type \"{}\" to \"{}\".",
+                const_expr.return_type(),
+                target
+            ))
```

Comment:
This error is now thrown all the way up to the user:
```
dev=> create table t (v1 int primary key);
CREATE_TABLE
dev=> select * from t where v1 < 1.5;
ERROR:  QueryError: Bind error: Cannot cast type "numeric" to "integer".
```

But it should just be a signal to skip the scan range optimization. Talking about its caller `analyze_eq_const_expr` behavior, it should do one of the 3 things:
* return `Ok(true)` so that the upper layer `return Ok(false_cond())`
* assign `eq_conds` and `return Ok(false)`
* signal the upper layer it [cannot match a known optimization](https://github.com/risingwavelabs/risingwave/blob/88dc35e44554152a88366f6535d7a60b37a4bf9f/src/frontend/src/utils/condition.rs#L467) and it should just `other_conds.push(expr);`

We are lacking the last possibility right now.